onnxruntime/core/providers/dml/dml_provider_factory.cc

```diff
@@ -216,6 +265,128 @@
+
+    // Used in the case that both GPU and NPU adapters are considered
+    ComPtr<IDXCoreAdapter>* gpu_adapters =
+        (ComPtr<IDXCoreAdapter>*) malloc(count * sizeof(ComPtr<IDXCoreAdapter>));
+    int gpu_adapter_count = 0;
+    ComPtr<IDXCoreAdapter>* npu_adapters =
+        (ComPtr<IDXCoreAdapter>*) malloc(count * sizeof(ComPtr<IDXCoreAdapter>));
+    int npu_adapter_count = 0;
+
+    // Iterate through all compute capable adapters
+    for (uint32_t i = 0; i < count; ++i)
+    {
+        ComPtr<IDXCoreAdapter> candidateAdapter;
+        ORT_THROW_IF_FAILED(
+            d3D12CoreComputeAdapters->GetAdapter(i, candidateAdapter.GetAddressOf()));
+
+        // Only considering hardware adapters
+        if (!isHardwareAdapter(candidateAdapter))
+            continue;
+
+        if (f == OrtDmlDeviceFilter::Gpu) // consider GPUs only
+        {
+            if (supportsGraphics(candidateAdapter)) {
+                ordered_adapters[ordered_adapter_count] = candidateAdapter;
+                ordered_adapter_count++;
+            }
+        }
+        else if(f == OrtDmlDeviceFilter::Npu) // consider NPUs only
+        {
+            if (!supportsGraphics(candidateAdapter)) {
+                ordered_adapters[ordered_adapter_count] = candidateAdapter;
+                ordered_adapter_count++;
+            }
+        }
+        else // consider both GPUs and NPUs
+        {
+            if (supportsGraphics(candidateAdapter)) {
+                gpu_adapters[gpu_adapter_count] = candidateAdapter;
+                gpu_adapter_count++;
+            } else {
+                npu_adapters[npu_adapter_count] = candidateAdapter;
+                npu_adapter_count++;
+            }
+        }
+    }
+
+    if (f == OrtDmlDeviceFilter::None) // considering both NPUs and GPUs
+    {
+        // order the adapters based on the performance preference
+        ComPtr<IDXCoreAdapter>* high_pri_adapters;
+        ComPtr<IDXCoreAdapter>* low_pri_adapters;
+        int num_high_pri_adapters;
+        int num_low_pri_adapters;
+        if (p == OrtDmlPerformancePreference::LowPower) // NPUs first
+        {
+            high_pri_adapters = npu_adapters;
+            num_high_pri_adapters = npu_adapter_count;
+            low_pri_adapters = gpu_adapters;
+            num_low_pri_adapters = gpu_adapter_count;
+        }
+        else // GPUs first
+        {
+            high_pri_adapters = gpu_adapters;
+            num_high_pri_adapters = gpu_adapter_count;
+            low_pri_adapters = npu_adapters;
+            num_low_pri_adapters = npu_adapter_count;
+        }
+
+        for (int i = 0; i < num_high_pri_adapters; i++) {
+            ordered_adapters[ordered_adapter_count] = high_pri_adapters[i];
+            ordered_adapter_count++;
+        }
+        for (int i = 0; i < num_low_pri_adapters; i++) {
+            ordered_adapters[ordered_adapter_count] = low_pri_adapters[i];
+            ordered_adapter_count++;
+        }
+    }
```

Comment:
The logic in this function can be summarized as:

if you have a low power filter,
   use a policy that puts NPUs first
if you have a high perf filter,
   use a policy that puts GPUs first
else do nothing

So....

if (filter == OrtDmlDeviceFilter::None) // considering both NPUs and GPUs
{
     struct SortingPolicy
     {
          bool gpusFirst_ = true;
          SortingPolicy(bool gpusFirst = true) : gpusFirst_(gpusFirst){}
          bool operator()(const AdapterInfo& a, const AdapterInfo& b) {
               return gpusFirst_ ? a.Type > b.Type : a.Type < b.Type;
          } 
     };

     auto sortingPolicy = SortingPolicy(preference != OrtDmlPerformancePreference::LowPower);
     std::sort(selected_adapters.begin(), selected_adapters.end(), sortingPolicy);
}